src/ReceiverLink.cs

```diff
@@ -351,10 +401,24 @@ internal Message ReceiveInternal(MessageCallback callback, int timeout = 60000)
         void DisposeMessage(Message message, Outcome outcome)
         {
             Delivery delivery = message.Delivery;
-            if (this.totalCredit > 0 &&
-            Interlocked.Increment(ref this.restored) >= (this.totalCredit / 2))
+            bool issueFlow = false;
+            int flowCredit = 0;
+            int dcSnd = 0;
+            lock (ThisLock)
+            {
+                Interlocked.Increment(ref deliveryCountRcv);
+                if (autoRestore && credit > 0 && restoreCountRcv == deliveryCountRcv)
+                {
+                    issueFlow = true;
+                    flowCredit = ComputeCredit(deliveryCountRcv, deliveryCountSnd, credit);
+                    dcSnd = deliveryCountSnd;
```

Comment:
dcSnd should probably be set a line earlier and passed to ComputeCredit instead of deliveryCountSnd to ensure the flow and credit calculation use a consistent value.

Reply:
Agreed. With these changes the self tests still pass!